math/minuit2/src/Numerical2PGradientCalculator.cxx

```diff
@@ -182,14 +167,16 @@ FunctionGradient Numerical2PGradientCalculator::operator()(const MinimumParamete
          grd(i) = 0.5*(fs1 - fs2)/step;
          g2(i) = (fs1 + fs2 - 2.*fcnmin)/step/step;
 
-#ifdef DEBUG
-         pr = std::cout.precision(13);
-         std::cout << "cycle " << j << " x " << x(i) << " step " << step << " f1 " << fs1 << " f2 " << fs2
-                   << " grd " << grd(i) << " g2 " << g2(i) << std::endl;
-         std::cout.precision(pr);
+#ifndef _OPENMP
```

Comment:
It would be nice to have these lines also availablewhen _OPENMP is enabled.
I was trying to debug the failing tests seen in OPENMP
Unfortunately if enabled, it is still crashing in MnPrint.Debug as before.
Suggested changes to reproduce it and improve printout quality
```
#ifdef _OPENMP
#pragma omp critical
#endif
         {
            if (i == 0 && j == 0) {
               print.Debug([&](std::ostream &os) {
                  os << std::setw(10) << "parameter" << std::setw(6) << "cycle" << std::setw(15) << "x" << std::setw(15)
                     << "step" << std::setw(15) << "f1" << std::setw(15) << "f2" << std::setw(15) << "grd"
                     << std::setw(15) << "g2" << std::endl;
               });
            }
            print.Debug([&](std::ostream &os) {
               const int pr = os.precision(13);
               const int iext = Trafo().ExtOfInt(i);
               os << std::setw(10) << Trafo().Name(iext) << std::setw(5) << j << "  " << x(i) << " " << step << " "
                  << fs1 << " " << fs2 << " " << grd(i) << " " << g2(i) << std::endl;
               os.precision(pr);
            });
         }
```

Reply:
If you want to print inside the thread, you need to create a new MnPrint instance inside the thread. Then there should be no crash. You can create one inside the braced scope { } here.

Reply:
I added the code as described.

Reply:
The core problem is really the use of global variables with threads. The is no golden way and no simple solution, we don't know where the thread is formed. MnPrint should work if a whole MnApplication is run in parallel and it should also work when a thread is started locally in Numerical2PCalculator. It is difficult to design something that works in both cases.